Content/WorldGeneration/DungeonGen/DungeonMaker.cs

```diff
@@ -0,0 +1,389 @@
+		/// cannot be made.
+		/// </summary>
+		/// <param name="x"></param>
+		/// <param name="y"></param>
+		/// <param name="directionDisposition"></param>
+		/// <returns></returns>
+		private bool TryMakeHallway(int x, int y, ref List<Point16> hallway, int remainingSections, Point16 directionDisposition = default)
+		{
+			if (hallway is null)
+				hallway = new List<Point16>();
+
+			hallway.Add(new Point16(x, y));
+
+			if (remainingSections <= 0)
+			{
+				AddHallway(hallway);
+				return true;
+			}
+
+			if (directionDisposition == default || WorldGen.genRand.NextBool(HALLWAY_TURN_DENOMINATOR))
+				directionDisposition = MakeRandomDirectionalPoint16();
+
+			for (int k = 0; k < MAX_HALLWAY_RETRIES; k++)
+			{
+				if (IsSectionValid(x + directionDisposition.X, y + directionDisposition.Y))
+					return TryMakeHallway(x + directionDisposition.X, y + directionDisposition.Y, ref hallway, remainingSections - 1, directionDisposition);
+
+				directionDisposition = MakeRandomDirectionalPoint16();
+			}
+
+			return false;
+		}
+
+		/// <summary>
+		/// Places a hallway into the dungeon.
+		/// </summary>
+		/// <param name="hallway"></param>
+		private void AddHallway(List<Point16> hallway)
+		{
+			for(int k = 1; k < hallway.Count - 1; k++)
+			{
+				var hall = hallway[k];
+				dungeon[hall.X, hall.Y] = secType.hall;
+				hallSections.Add(hall);
+			}
+		}
+
+		/// <summary>
+		/// Removes a hallway from the dungeon.
+		/// </summary>
+		/// <param name="hallway"></param>
+		private void RemoveHallway(List<Point16> hallway)
+		{
+			for (int k = 1; k < hallway.Count - 1; k++)
+			{
+				var hall = hallway[k];
+				dungeon[hall.X, hall.Y] = secType.none;
+				hallSections.Remove(hall);
+			}
+		}
+
+		/// <summary>
+		/// Returns a random cardinal direction Point16. Used by the hallway generator to create turns.
+		/// </summary>
+		/// <returns></returns>
+		private Point16 MakeRandomDirectionalPoint16()
+		{
+			var rand = WorldGen.genRand.Next(4);
+
+			switch (rand)
+			{
+				case 0: return new Point16(0, 1);
+				case 1: return new Point16(0, -1);
+				case 2: return new Point16(1, 0);
+				case 3: return new Point16(-1, 0);
+				default: return new Point16(1, 0);
+			}
+		}
+
+		/// <summary>
+		/// Generates a "limb" of the dungeon recursively. This attempts to generate a hallway with a room attached
+		/// to the end. If it is unable to place a room, it will backpedal and destroy the hallway. It will also
+		/// attempt to "branch" off of each room it makes if possible and depending on the branch chance.
+		/// </summary>
+		/// <param name="start"></param>
+		/// <param name="remainingRooms"></param>
+		private void GenerateLimb(Point16 start, int remainingRooms)
+		{
+			if (remainingRooms <= 0)
+				return;
+
+			List<Point16> hall = new List<Point16>();
+			if (TryMakeHallway(start.X, start.Y, ref hall, WorldGen.genRand.Next(3, 7)))
+			{
+				if (TryMakeRandomRoom(hall.Last().X, hall.Last().Y))
```

Comment:
I think TryMakeRandomRoom should pass in remainingRooms, so that we can choose to have rooms near the "beginning" of the dungeon (but not necessarily at the start)